---
title: "Fig.S6b & Fig.S6c differential genes and taxanomy contribution of mgPGPT"
author: "Huanhuan Zhu"
date: "2025-10-09"
output: html_document
---


## Fig.S6b Differentail genes
```{r, fig.align='center', fig.height=6, fig.width=6.3, warning=FALSE, message=FALSE}

library(tidyverse)
library(ggplot2)
library(tibble)
library(edgeR)
library(data.table)
library(patchwork)
library(ggh4x)
library(scales)
library(dplyr)
library(forcats)

rm(list=ls())

setwd("E:/HMME/Metagenome/Result_new/mgPGPT/")

ko_all <- read.table("mgPGPT.summarizeAbundance.count.sseqid.raw.txt",
                     check.names = F,
                     row.names = 1,
                     header = T)
#ko_all <- ko_all[2:nrow(ko_all),]
ko_all_up <- ko_all
ko_all_up <- ko_all_up[,grepl("TS",colnames(ko_all_up))]

set.seed(123)
count_data <- ko_all_up

count_data <- round(count_data)


col_data <- data.frame(
  sample = colnames(count_data),
  drought = as.numeric(substr(colnames(count_data),1,1))
)
rownames(col_data) <- col_data$sample


# edger
dge <- DGEList(counts = count_data, group = col_data$drought)

keep <- filterByExpr(dge)

dge <- dge[keep,, keep.lib.sizes=FALSE]

dge <- calcNormFactors(dge, method = "TMM")

cpm_norm <- cpm(dge, normalized.lib.sizes = TRUE)

plotMDS(dge, dim = c(1, 2))

design <- model.matrix(~ col_data$drought)

dge <- estimateDisp(dge, design, robust = TRUE) 

plotBCV(dge)

fit <- glmFit(dge, design, robust = TRUE)     #????циб??
lrt <- glmLRT(fit)

res <- topTags(lrt, n = Inf)
res_df <- res$table
res_df$KO <- rownames(res_df)

res_sig <- res_df[res_df$FDR < 0.05, ]

res_sig$group <- "Up"
res_sig$group[res_sig$logFC < 0] <- "Down"

pathway <- readRDS("E:/Methods/plabase/PlaBAse.rds")

res_sig <- left_join(res_sig, pathway, by=c("KO"="ID"))
res_sig$Lv3_rename <- gsub("\\|","_",res_sig$Lv3)
res_sig$id <- paste0(res_sig$group, "_", res_sig$Lv3_rename)

for (i in unique(res_sig$id)){
  res_sig_tmp <- res_sig[res_sig$id == i,c("KO")]
  write.table(res_sig_tmp, paste0("E:/HMME/Metagenome/Result_new/mgPGPT/group/",i,".txt"), row.names = FALSE, quote = FALSE, col.names = FALSE)
}

pathway1 <- pathway
pathway1$Lv3 <- gsub("\\|","_",pathway1$Lv3)
for (i in unique(pathway1$Lv3)){
  res_sig_tmp <- pathway1[pathway1$Lv3 == i,c("ID")]
  write.table(res_sig_tmp, paste0("E:/HMME/Metagenome/Result_new/mgPGPT/group_all/",i,".txt"), row.names = FALSE, quote = FALSE, col.names = FALSE)
}


res_df$group <- "Not significant"
res_df$group[res_df$logFC > 0 & res_df$FDR < 0.05] <- "Up"
res_df$group[res_df$logFC < 0 & res_df$FDR < 0.05] <- "Down"


tbl <- table(res_df$group)

txt <- paste0(
  "Up = ", tbl["Up"], "\n",
  "Down = ", tbl["Down"], "\n",
  "Not significant = ", tbl["Not significant"]
)


res_df$x = "1"
set.seed(123)
p <- ggplot(res_df, aes(x = x, y = logFC, color = factor(group, levels = c("Up", "Down", "Not significant")))) +
  geom_point(alpha = 0.6, size = 2, position = position_jitter(width = 0.5)) +
  scale_color_manual(name="Type", 
                     values = c("Up" = "red", "Down" = "blue", "Not significant" = "grey"),
                     labels =c("Up" = "Up (n = 1755)", "Down" = "Down (n = 1659)", "Not significant" = "Not significnat (n = 2138)"),
                     guide = guide_legend(nrow=3,
                     override.aes = list(size = 6))) +
  theme_bw() +
  # annotate("text", x= 1,y=3,   label = txt,
  #          hjust = 0, vjust = 1,
  #          size = 8) +
  labs(x = NULL, y = "Log2 Fold Change", title = "Differential PGPT genes") +
  #geom_vline(xintercept = c(-1, 1), linetype = "dashed") +
  #geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  theme(plot.title = element_text(size=30, hjust=0.5),
        axis.title = element_text(size=26),
        axis.text.y = element_text(size=22, color="black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = "bottom",
        legend.title = element_text(size=26),
        legend.text = element_text(size=22))


#w635 h600
print(p)

```


## Fig.S6b Taxonomy contribution on increased genes
```{r, fig.align='center', fig.height=22, fig.width=16, warning=FALSE, message=FALSE}


res_df <- left_join(res_df, pathway, by=c("KO"="ID"))



table(res_df[res_df$logFC > 0 & res_df$FDR < 0.05,]$Lv3)

test1 <- as.data.frame(table(res_df[res_df$logFC > 0 & res_df$FDR < 0.05,]$Lv3))
test <- as.data.frame(table(res_df[res_df$logFC < -0.5 & res_df$FDR < 0.05,]$Lv3))


cpm_norm2 <- as.data.frame(cpm_norm[res_sig$KO,]) 
cpm_norm2$ID <- row.names(cpm_norm2)

cpm_norm2 <- left_join(cpm_norm2, res_sig[,c("group", "KO", "Lv3")], by=c("ID" = "KO"))
cpm_norm2 <- aggregate(.~group+Lv3, cpm_norm2[,c(1:35,37:38)], FUN="sum")


sample_cols <- grep("\\.quant$", names(cpm_norm2), value = TRUE)

all_combos <- expand.grid(
  Lv3 = unique(cpm_norm2$Lv3),
  group = c("Up", "Down"),
  stringsAsFactors = FALSE
)

df_full <- all_combos %>%
  left_join(cpm_norm2, by = c("Lv3", "group")) %>%
  mutate(across(all_of(sample_cols), ~replace_na(., 0)))

df_full


cpm_norm2 <- df_full

cpm_norm2 <- reshape2::melt(cpm_norm2, id=c("group","Lv3"))
cpm_norm2$variable <- substr(cpm_norm2$variable, 1, 1)


up_lv3 <- cpm_norm2 %>% filter(group == "Up") %>% pull(Lv3)
down_lv3 <- cpm_norm2 %>% filter(group == "Down") %>% pull(Lv3)

common_lv3 <- intersect(up_lv3, down_lv3)
diff_lv3 <- setdiff(c(up_lv3, down_lv3), intersect(up_lv3, down_lv3))

df <- NULL
for (i in unique(common_lv3)){
  result <- t.test(cpm_norm2[cpm_norm2$Lv3 == i & cpm_norm2$group == "Up",]$value, 
                   cpm_norm2[cpm_norm2$Lv3 == i & cpm_norm2$group == "Down",]$value)
  df_tmp <- data.frame("name"=i, "p"=result$p.value, "value" = (mean(cpm_norm2[cpm_norm2$Lv3 == i & cpm_norm2$group == "Up",]$value) -
                                                                  mean(cpm_norm2[cpm_norm2$Lv3 == i & cpm_norm2$group == "Down",]$value)))
  df <- rbind(df, df_tmp)
}

df$p <- p.adjust(df$p, method = "fdr")

df2 <- df

df2$type = "Up"
df2[df2$p > 0.05,]$type = "Stable"
df2[df2$value < 0,]$type = "Down"

df2 <- df2[,c("name","type")]

##df_no < data.frame("name"= diff_lv3, "type" = "Stable")
##df2<- bind(df2, df_no)
df2$name <- tolower(df2$name)
df2$type <- factor(df2$type, levels = c("Up","Down","Stable"))
cpm_norm2 <- aggregate(value~group+Lv3+variable, cpm_norm2, FUN="mean")


pathway2 <- pathway[,c("Lv2", "Lv3")]
pathway2 <- pathway2[!duplicated(pathway2$Lv3),]

cpm_norm2 <- left_join(cpm_norm2, pathway2, by="Lv3")

cpm_norm2$Lv2 <- tolower(cpm_norm2$Lv2)
cpm_norm2$Lv3 <- tolower(cpm_norm2$Lv3)

cpm_norm2 <- cpm_norm2 %>%
  group_by(Lv3) %>%
  mutate(value_scaled = as.numeric(scale(value))) %>%
  ungroup()

order_new <- aggregate(value_scaled~Lv3+Lv2, cpm_norm2[cpm_norm2$group =="Up",], FUN="mean")



order_new <- order_new %>%
  arrange(Lv2, desc(value_scaled)) %>%
  mutate(Lv3 = factor(Lv3, levels = unique(Lv3)))

# taxonomy
folder_path <- "E:/HMME/Metagenome/Result_new/mgPGPT/group_result/family/"

all_files <- list.files(folder_path, pattern = "\\.txt$", full.names = TRUE)

result_list <- list()

for (file in all_files) {
  
  test <- as.data.frame(fread(file))
  test <- test[test[,1] != "Others", ]
  
  rownames(test) <- NULL
  type <- colnames(test)[1]
  test <- column_to_rownames(test, type)
  test <- test[, grepl("TS", colnames(test))]
  
  test <- as.data.frame(apply(test, 2, function(x) x / sum(x) * 100))
  
  test$id <- rownames(test)
  test <- reshape2::melt(test, "id")
  test$variable <- substr(test$variable,1,1)

  test_order <- aggregate(value ~ id, test[test$variable == "6",], FUN = "mean")
  #test_order <- test_order[test_order$id %in% order_final, ]
  test_order <- test_order[order(test_order$value, decreasing = TRUE), ]
  
  test[!(test$id %in% test_order$id[1:20]), "id"] <- "Others"
  
  test$type <- tolower(type)

  test$file <- basename(file)
  
  result_list[[file]] <- test
}

test_all <- bind_rows(result_list)

order_first <- as.data.frame(table(test_all$id))
order_first <- order_first[order(order_first$Freq, decreasing = T), ]
order_final <- as.character(order_first$Var1[3:22])

result_list <- list()

for (file in all_files) {
  
  test <- as.data.frame(fread(file))
  test <- test[test[,1] != "Others", ]
  
  rownames(test) <- NULL
  type <- colnames(test)[1]
  test <- column_to_rownames(test, type)
  test <- test[, grepl("TS", colnames(test))]
  
  test <- as.data.frame(apply(test, 2, function(x) x / sum(x) * 100))
  
  test$id <- rownames(test)
  test <- reshape2::melt(test, "id")
  test$variable <- substr(test$variable,1,1)
  
  test_order <- aggregate(value ~ id, test[test$variable == "6",], FUN = "mean")
  test_order <- test_order[test_order$id %in% order_final, ]
  test_order <- test_order[order(test_order$value, decreasing = TRUE), ]
  
  test[!(test$id %in% test_order$id), "id"] <- "Others"
  
  test$type <- tolower(type)

  test$file <- basename(file)
  
  result_list[[file]] <- test
}


test_all <- bind_rows(result_list)

test_all$id <- sub(";$", "", test_all$id)
test_all$type <- sub(".*up_", "", test_all$type)
test_all <- aggregate(value~variable+id+type, test_all, FUN = "sum")
test_all$value <- test_all$value/5


family_order <- aggregate(value~id, test_all[test_all$variable =="6",], FUN = "sum")
family_order <- family_order[order(family_order$value, decreasing = T),]

pathway <- readRDS("E:/Methods/plabase/PlaBAse.rds")
pathway <- pathway[,8:9]
pathway <- pathway[!duplicated(pathway$Lv3),]
pathway$Lv2 <- tolower(pathway$Lv2)
pathway$Lv3 <- tolower(pathway$Lv3)

pathway$Lv3 <- gsub("\\|","_",pathway$Lv3)

pathway <- pathway[pathway$Lv3 %in% test_all$type,]
pathway <- pathway %>%
  arrange(Lv2, Lv3)

my_colors <- c(
  "Taxonomy contribution" = alpha("#1445f8", 1)
)

strip_custom <- strip_themed(
  background_x = elem_list_rect(
    fill = my_colors,
    color = "black"
  ),
  text_x = elem_list_text(
    #face = "bold",
    colour = "white"
  ))


order_new2 <- gsub("\\|","_",rev(c(order_new$Lv3)))
test_all$taxo <- "Taxonomy contribution"

colors <- c(
  "#FFD700", "#2196F3", "#8A4B08",
  "#FF7F00", "#86519D", "#1B9E77", 
  "#3B3B98", "#E41A1C")


main2 <- ggplot(test_all[test_all$variable =="6",], aes(x = factor(id, levels=c(family_order[family_order$id != "Others", ]$id,"Others")), y = factor(type, order_new2), fill = log(value+1))) +
  geom_tile(color = "white") +
  scale_fill_gradient2(name="Log abundance", low = "white", mid = "red", high = "darkred", midpoint = 2.0) +
  ggtitle("Taxonomy contribution") +
  theme_void() +
  theme(
    plot.title = element_text(size = 40, hjust = 0.5),
    axis.text.x = element_text(color="black",size=23, angle = 90, vjust = 0.5, hjust = 1, margin = margin(t=5)),
    axis.title = element_blank(),
    strip.text = element_text(size=31),
    axis.text.y = element_blank(),
    legend.key.height = unit(1, "cm"),
    legend.key.width = unit(0.6, "cm"),
    legend.title = element_text(size = 27),
    legend.text = element_text(size = 23)
  )
main2



function_group <- ggplot(cpm_norm2, aes(x = "", 
                                        y = factor(Lv3, levels = rev(c(order_new$Lv3))),color = Lv2)) +
  geom_point(size=9) +
  labs(y="Lv3") +
  scale_color_manual(name="mgPGPT level 2", values = colors, guide = guide_legend(size=5))+
  scale_x_discrete(expand = expansion(mult = c(0.01, 0.01))) +
  scale_y_discrete(position = "right") +
  theme_void() +
  theme( 
    axis.text.y = element_text(size=23, hjust = 0, margin = margin(l = 10,unit = "pt")),
    legend.title = element_text(size = 27),
    legend.text = element_text(size = 23),
    legend.key.height = unit(31, "pt"))
#+theme(axis.text.y = element_text(size=8, hjust = 1))

function_group


# 
p <- (main2|function_group) +plot_layout(guides = "collect",widths = c(35,1.5)) &
  theme(plot.margin = margin(r=10),
        legend.position = "bottom",  
        legend.direction="vertical",
        legend.key.spacing.y = unit(-8, "pt"),
        legend.box.margin = margin(l = 40, unit = "cm")) 

print(p)

# w22 16


```




