---
title: "Fig.4d & Fig.S7b_MAG_mgPGPT_boxplot"
author: "Huanhuan Zhu"
date: "2025-09-28"
output: html_document
---

## Fig. S7b_Gene frequency and genome size for MAG phylum
```{r, fig.align='center', fig.height=6.20, fig.width=16.5, warning=FALSE, message=FALSE}

library(ggh4x)
library(ggtree)
library(ape)
library(ggtreeExtra)
library(ggnewscale)
library(scales)
library(agricolae)
library(ggnewscale)
library(tidyverse)
library(data.table)
library(edgeR)

rm(list=ls())

mag_abundance <- read.table("E:/HMME/Metagenome/Result_new/bin/bin_dereplecated/coverm_prokaryotes/all_bin_coverages_TR_count.tsv",
                            sep = "\t", header = T, check.names = F, row.names = 1)

mag_abundance <- as.data.frame(t(mag_abundance))


metadata <- data.frame("id" = rownames(mag_abundance), "group" = substr(rownames(mag_abundance),1,1))
metadata <- column_to_rownames(metadata, "id")

mag_abundance <- as.data.frame(t(mag_abundance))


# edger for differential genes
dge <- DGEList(counts = mag_abundance, group = metadata$group)

keep <- filterByExpr(dge)
dge <- dge[keep,, keep.lib.sizes=FALSE]

dge <- calcNormFactors(dge, method = "TMM")

design <- model.matrix(~ metadata$group)

dge <- estimateDisp(dge, design, robust = TRUE) 


fit <- glmFit(dge, design, robust = TRUE)
lrt <- glmLRT(fit)

res <- topTags(lrt, n = Inf)
res_df <- res$table
res_df$KO <- rownames(res_df)

res_sig <- res_df[res_df$FDR < 0.05 , ]


mag_info <- read.table("E:/HMME/Metagenome/Result_new/bin/bin_dereplecated/checkm2/quality_report.tsv", sep = "\t", header = T)

mag_taxa <- read.table("E:/HMME/Metagenome/Result_new/bin/bin_dereplecated/gtdbtk_skip_ani/gtdbtk.bac120.summary.tsv", sep = "\t", header = T)
mag_taxa <- mag_taxa %>% 
  #filter(user_genome %in% mag_tree_filtered$tip.label,) %>% 
  separate(classification, sep = ";", into=c("kingdom","phylum","class","order","family", "genus", "species")) %>%
  select(user_genome,phylum)  %>%
  mutate("label"=user_genome)

bin_info <- left_join(mag_taxa, mag_info, by=c("user_genome"="Name"))
bin_info <- left_join(bin_info, res_df, by=c("user_genome"="KO"))

pgpt_filt <- read.csv("E:/HMME/Metagenome/Result_new/bin/bin_dereplecated/mgPGPT/drought_pgpt_filted.csv")


bin_pgpt <- read.table("E:/HMME/Metagenome/Result_new/bin/bin_dereplecated/mgPGPT/mgPGPT_default_filt.tsv", sep = "\t", check.names = F, header = F)
#bin_pgpt <- separate(bin_pgpt, "V2", sep = "_", into = "PGPT")


setDT(bin_pgpt)  # 转为 data.table
bin_pgpt[, PGPT := tstrsplit(V2, "_", keep = 1)]
bin_pgpt$ID <- sub("_[^_]+$", "", bin_pgpt$V1)


bin_info <- bin_info[bin_info$Completeness > 90 & bin_info$Contamination < 5, ]
#bin_info <- bin_info[(bin_info$logFC > 0 & bin_info$FDR < 0.05),]

bin_info <- separate(bin_info, phylum, sep = "__", into = c("pre_phylum", "Phylum"))

bin_info1 <- bin_info[bin_info$logFC > 0 & bin_info$FDR < 0.05, ]
selected <- as.data.frame(table(bin_info1$Phylum))
selected1 <- as.character(selected[selected$Freq >= 5,]$Var1)


# pgpt_richess
pgpt_richess <- NULL
for (i in bin_info[bin_info$logFC > 0 & bin_info$FDR < 0.05, ]$user_genome){
  
  richness_tmp <- length(unique(bin_pgpt[bin_pgpt$ID == i,]$PGPT))
  df_tmp <- data.frame("label" = i, "Phylum" = bin_info[bin_info$user_genome == i,]$Phylum, "Richness" = richness_tmp )
  pgpt_richess <- rbind(pgpt_richess, df_tmp)
}


# bin treads
bin_trend <- bin_info[, c("label","logFC","FDR")]
bin_trend$type = "Stable"
bin_trend[bin_trend$logFC < 0 & bin_trend$FDR <0.05, ]$type = "Decrease"
bin_trend[bin_trend$logFC > 0 & bin_trend$FDR <0.05, ]$type = "Increase"
bin_trend$type <- factor(bin_trend$type, levels=c("Increase","Decrease","Stable"))


# bin drought_associated 
pgpt_filt <- read.csv("E:/HMME/Metagenome/Result_new/bin/bin_dereplecated/mgPGPT/drought_pgpt_filted.csv")
pathway <- readRDS("E:/Methods/plabase/PlaBAse.rds")


# pgpt classification
osm <- pgpt_filt[pgpt_filt$type2 == "OSMOTIC or SALINITY_STRESS",]$PGPT  # Lv5
biolfilm <- pgpt_filt[pgpt_filt$type2 == "Biofilm",]$PGPT  # Lv5
spore <- pgpt_filt[pgpt_filt$type2 == "spore",]$PGPT  # Lv5
ros <- pgpt_filt[pgpt_filt$type2 == "ROS",]$PGPT  # Lv5
eps <- pgpt_filt[pgpt_filt$type2 == "eps",]$PGPT  # Lv3
eps <- unique(pathway[pathway$Lv3 == eps,]$Lv5)


C <- pgpt_filt[pgpt_filt$type2 == "C",]$PGPT  # Lv3
N <- pgpt_filt[pgpt_filt$type2 == "N",]$PGPT  # Lv3
P <- pgpt_filt[pgpt_filt$type2 == "P",]$PGPT  # Lv3
S <- pgpt_filt[pgpt_filt$type2 == "S",]$PGPT  # Lv3
K <- pgpt_filt[pgpt_filt$type2 == "K",]$PGPT  # Lv3
siderophores <- pgpt_filt[pgpt_filt$type2 == "siderophores",]$PGPT  # Lv4
phytohormone <- pgpt_filt[pgpt_filt$type2 == "phytohormone",]$PGPT  # Lv4
colonization <-  pgpt_filt[pgpt_filt$type2 == "colonization",]$PGPT  # Lv3

drought <- c(osm, biolfilm, spore, ros, eps)
drought_pgpt <- unique(pathway[pathway$Lv5 %in% drought,]$ID)


pgpt_drought <- NULL
for (i in bin_info[bin_info$logFC > 0 & bin_info$FDR < 0.05, ]$user_genome){
  
  drought_tmp <- table(bin_pgpt[bin_pgpt$ID == i,]$PGPT %in% drought_pgpt)
  df_tmp <- data.frame("label" = i, "Phylum" = bin_info[bin_info$user_genome == i,]$Phylum, "Drought" = as.numeric(drought_tmp["TRUE"]) )
  pgpt_drought <- rbind(pgpt_drought, df_tmp)
}


growth_pgpt <- c(C,N,P,S,K,siderophores,phytohormone,colonization)

growth_pgpt3 <- unique(pathway[pathway$Lv3 %in% growth_pgpt,]$ID)
growth_pgpt4 <- unique(pathway[pathway$Lv4 %in% growth_pgpt,]$ID)

growth_pgpt <- unique(growth_pgpt3,growth_pgpt4)


pgpt_growth <- NULL
for (i in bin_info[bin_info$logFC > 0 & bin_info$FDR < 0.05, ]$user_genome){
  
  drought_tmp <- table(bin_pgpt[bin_pgpt$ID == i,]$PGPT %in% growth_pgpt)
  df_tmp <- data.frame("label" = i, "Phylum" = bin_info[bin_info$user_genome == i,]$Phylum, "Growth" = as.numeric(drought_tmp["TRUE"]) )
  pgpt_growth <- rbind(pgpt_growth, df_tmp)
}


species_size_plot <- mag_info%>%select(Name, Genome_Size)%>%mutate(Genome_Size2=Genome_Size/1000000)
species_size_plot <- na.omit(left_join(species_size_plot, bin_info[bin_info$logFC > 0 & bin_info$FDR < 0.05, ][,c("user_genome", "Phylum")], by=c("Name" = "user_genome")))
names(species_size_plot)[1] = "label"

# richness
compare_richness <- with(pgpt_richess[pgpt_richess$Phylum %in% selected1, ],kruskal(Richness,Phylum,group = TRUE, p.adj = "fdr"))
compare_richness
compare_richness1 <- compare_richness$groups
compare_richness1$Phylum <- rownames(compare_richness1)
compare_richness1 <- compare_richness1[,2:3]
colnames(compare_richness1)[1] = "Richness"

compare_richness_max <- compare_richness$means
compare_richness_max$Phylum <- rownames(compare_richness_max)
compare_richness1 <- left_join(compare_richness1, compare_richness_max[,c("Phylum","Max")], by = "Phylum")

# richness_order
compare_richness_order <- with(pgpt_richess,kruskal(Richness,Phylum,group = TRUE, p.adj = "fdr"))
compare_richness_order

# drought
compare_drought <- with(pgpt_drought[pgpt_drought$Phylum %in% selected1, ],kruskal(Drought,Phylum,group = TRUE, p.adj = "fdr"))
compare_drought
compare_drought1 <- compare_drought$groups
compare_drought1$Phylum <- rownames(compare_drought1)
compare_drought1 <- compare_drought1[,2:3]
colnames(compare_drought1)[1] = "Drought"

compare_drought_max <- compare_drought$means
compare_drought_max$Phylum <- rownames(compare_drought_max)
compare_drought1 <- left_join(compare_drought1, compare_drought_max[,c("Phylum","Max")], by = "Phylum")

# growth
compare_growth <- with(pgpt_growth[pgpt_growth$Phylum %in% selected1, ],kruskal(Growth,Phylum,group = TRUE, p.adj = "fdr"))
compare_growth
compare_growth1 <- compare_growth$groups
compare_growth1$Phylum <- rownames(compare_growth1)
compare_growth1 <- compare_growth1[,2:3]
colnames(compare_growth1)[1] = "Growth"

compare_growth_max <- compare_growth$means
compare_growth_max$Phylum <- rownames(compare_growth_max)
compare_growth1 <- left_join(compare_growth1, compare_growth_max[,c("Phylum","Max")], by = "Phylum")


# genome size
compare_gs <- with(species_size_plot[species_size_plot$Phylum %in% selected1, ],kruskal(Genome_Size2,Phylum,group = TRUE, p.adj = "fdr"))
compare_gs
compare_gs1 <- compare_gs$groups
compare_gs1$Phylum <- rownames(compare_gs1)
compare_gs1 <- compare_gs1[,2:3]
colnames(compare_gs1)[1] = "Genome_Size2"

compare_gs_max <- compare_gs$means
compare_gs_max$Phylum <- rownames(compare_gs_max)
compare_gs1 <- left_join(compare_gs1, compare_gs_max[,c("Phylum","Max")], by = "Phylum")


# plot df
list(pgpt_richess, pgpt_drought[,c(1,3)], pgpt_growth[,c(1,3)], species_size_plot[,c(1,3)]) %>%
  purrr::reduce(full_join, by = "label") %>%
  reshape2::melt(c("label", "Phylum")) -> pgpt_all

# significance df
list(compare_richness1, compare_drought1, compare_growth1, compare_gs1) %>%
  purrr::reduce(full_join, by = "Phylum") %>%
  dplyr::rename(
    Richness_max = Max.x,
    Drought_max = Max.y,
    Growth_max = Max.x.x,
    Genome_Size2_max = Max.y.y
  ) %>%
  pivot_longer(
    cols = c(Richness, Drought, Growth, Genome_Size2),
    names_to = "variable",
    values_to = "value"
  ) %>%
  pivot_longer(
    cols = ends_with("_max"),
    names_to = "temp",
    values_to = "max"
  ) %>%
  mutate(temp = gsub("_max", "", temp)) %>%
  filter(temp == variable) %>%
  select(Phylum, variable, value, max) -> compare_all

compare_all[compare_all$variable == 'Richness',]$max <- compare_all[compare_all$variable == 'Richness',]$max + 200
compare_all[compare_all$variable == 'Growth',]$max <- compare_all[compare_all$variable == 'Growth',]$max + 400
compare_all[compare_all$variable == 'Drought',]$max <- compare_all[compare_all$variable == 'Drought',]$max + 100
compare_all[compare_all$variable == 'Genome_Size2',]$max <- compare_all[compare_all$variable == 'Genome_Size2',]$max + 1

compare_all$variable <- factor(compare_all$variable, levels=c("Richness", "Drought", "Growth", "Genome_Size2"))


# phylum counts
phylum_count <- as.data.frame(table(bin_info[bin_info$logFC > 0 & bin_info$FDR < 0.05, ]$Phylum))
phylum_count$new_phylum <- paste0(phylum_count$Var1," (n=", phylum_count$Freq, ")")
phylum_labels <- setNames(phylum_count$new_phylum, phylum_count$Var1)


# three
# strip color
my_colors <- c(
  "Drought" = alpha("#A56B3F", 1),
  "Growth" = alpha("#3D7E0F", 1),
  "Genome_Size2" = alpha("#4D4D4D", 1)
)

strip_custom <- strip_themed(
  background_x = elem_list_rect(
    fill = my_colors,
    color = "black"
  ),
  text_x = elem_list_text(
    #face = "bold",
    colour = "white"
  )
)

p <- ggplot(pgpt_all[pgpt_all$variable != "Richness", ], aes(y = factor(Phylum, levels=rev(rownames(compare_richness_order$means)[order(compare_richness_order$means$rank, decreasing = T)])),
                     x = value, color = factor(variable))) + 
  facet_wrap2(.~variable, strip = strip_custom, scales = "free_x", ncol=4,   labeller = labeller(
    variable = c(
      "Richness" = " Richness",
      "Drought" = "Drought",
      "Growth" = "Growth",
      "Genome_Size2" = "Genome size (Mb)"
    ))) +
  geom_boxplot(width = 0.6, outlier.size = 1) +
  geom_point(position = position_jitter(height = 0.2), size = 0.1, alpha = 0.8) +
  geom_text(data = compare_all[compare_all$variable != "Richness", ], aes(y = Phylum, x = max, label = value), color = "black", size = 8)+
  labs(x = NULL, y="Phylum (Increased MAGs)") +
  scale_y_discrete(labels = phylum_labels, position = "right") +
  scale_color_manual(name = NULL, values = c(
    "Richness" ="#CB181D",
    "Drought" ="#A56B3F",
    "Growth" = "#3D7E0F",
    "Genome_Size2" = "#4D4D4D"),
    guide = "none"
  ) +
  theme_bw() + 
  theme(
    axis.text.y = element_text(size = 22, color = "black"),
    axis.text.x = element_text(size = 22, angle = 45, color = "black", hjust = 1),
    strip.text = element_text(size = 26),
    axis.title = element_text(size = 26)
  )

print(p)
# 1650 620
```
  
  
## Fig. 4d_mgPGPTrichness for MAG phylum
```{r, fig.align='center', fig.height=7.25, fig.width=9.5, warning=FALSE, message=FALSE}  
  #only richness
  
  pgp_gs <- pgpt_all[pgpt_all$variable == "Genome_Size2",]
  
  # genome_size added
  mean_gs <- aggregate(value~ Phylum, pgp_gs, FUN="mean")
  colnames(mean_gs)[2] <- "mean"
  
  sd_gs <- aggregate(value~ Phylum, pgp_gs, FUN="sd")
  colnames(sd_gs)[2] <- "sd"
  
  all_gs <- merge(mean_gs, sd_gs, by = "Phylum")
  
  all_gs$all <- paste0(all_gs$Phylum,
                       " (", round(all_gs$mean,2), " ± ",  round(all_gs$sd,2), " Mb)")
  
  
  phylum_count_gs <- merge(phylum_count, all_gs, by.x = "Var1", by.y = "Phylum")
  
  phylum_count_gs$label <- paste0(
    phylum_count_gs$Var1,
    " (n=", phylum_count_gs$Freq,
    ", ", round(phylum_count_gs$mean, 2),
    " ± ", round(phylum_count_gs$sd, 2), " Mb)"
  )
  
  
  
  phylum_count_gs$label2 <- paste0(
    " n = ", phylum_count_gs$Freq,
    ", ", round(phylum_count_gs$mean, 2),
    " ± ", round(phylum_count_gs$sd, 2), " Mb"
  )

  
  phylum_labels <- setNames(phylum_count_gs$label, phylum_count_gs$Var1)
  
  phylum_count_gs$x <-  -2000
  
  p <- ggplot(pgpt_all[pgpt_all$variable == "Richness",], aes(y = factor(Phylum, levels=rev(rownames(compare_richness_order$means)[order(compare_richness_order$means$rank, decreasing = T)])),
                                                         x = value, color = factor(variable))) + 
    geom_boxplot(width = 0.6, outlier.size = 1) +
    geom_point(position = position_jitter(height = 0.2), size = 0.1, alpha = 0.8) +
    geom_text(data = compare_all[compare_all$variable =="Richness",], aes(y = Phylum, x = max+50, label = value), color = "black", size=7)+
    geom_text(data = phylum_count_gs, aes(x=x, y=Var1, label = label2), color = "black", size = 7, hjust=0) +
    labs(x = NULL, y="MAG Phylum", title = "Gene richness of\n increased MAGs") +
    #scale_y_discrete(labels = phylum_labels, position = "left") +
    scale_color_manual(name = NULL, values = c(
      "Richness" ="#CB181D",
      "Drought" ="#A56B3F",
      "Growth" = "#3D7E0F",
      "Genome_Size2" = "#4D4D4D"),
      guide = "none"
    ) +
    scale_x_continuous(
      breaks = c( 500, 1000, 1500, 2000),
      limits = c(-2000, 2500)
    ) +
    theme_bw() + 
    theme(
      plot.title = element_text(size=36, hjust=0.5),
      axis.text.y = element_text(size = 24, color = "black"),
      axis.text.x = element_text(size = 24, angle = 45, color = "black", hjust = 1, vjust = 1.0),
      strip.text = element_text(size = 22),
      axis.title = element_text(size = 30)
    )
  
  print(p)
  # 900 height725
  
```
