---
title: "Fig.S4b & Fig.S4d_Differential genes and enrichenemnt analysis of KEGG"
author: "Huanhuan Zhu"
date: "2025-10-09"
output: html_document
---


## Fig.S4b Differential genes
```{r, fig.align='center', fig.height=6.35, fig.width=6.0, warning=FALSE, message=FALSE}

#############################Fig.S5b & Fig.S5b_Differential genes and enrichenemnt analysis of KEGG##########################

library(DESeq2)
library(ggplot2)
library(tibble)
library(edgeR)
library(clusterProfiler)
library(ggh4x) # different limit for each facet

rm(list=ls())

setwd("E:/HMME//Metagenome/Result_new/eggnog/")

ko_all <- read.table("eggnog.summarizeAbundance.count.KEGG_ko.raw.txt",
                     check.names = F,
                     row.names = 1,
                     header = T)
ko_all <- ko_all[2:nrow(ko_all),]
ko_all_up <- ko_all
ko_all_up <- ko_all_up[,grepl("TS",colnames(ko_all_up))]


set.seed(123)
count_data <- ko_all_up
#count_data <- count_data[,grepl("TS",colnames(count_data))]

count_data <- round(count_data)

col_data <- data.frame(
  sample = colnames(count_data),
  drought = as.numeric(substr(colnames(count_data),1,1))
)
rownames(col_data) <- col_data$sample


dge <- DGEList(counts = count_data, group = col_data$drought)

keep <- filterByExpr(dge)
dge <- dge[keep,, keep.lib.sizes=FALSE]

dge <- calcNormFactors(dge, method = "TMM")

cpm_norm <- cpm(dge, normalized.lib.sizes = TRUE)

plotMDS(dge, dim = c(1, 2))

design <- model.matrix(~ col_data$drought)

dge <- estimateDisp(dge, design, robust = TRUE)

plotBCV(dge)

fit <- glmFit(dge, design, robust = TRUE)
lrt <- glmLRT(fit)

res <- topTags(lrt, n = Inf)
res_df <- res$table
res_df$KO <- rownames(res_df)

res_sig <- res_df[res_df$FDR < 0.05, ]

res_df$group <- "Not significant"
res_df$group[res_df$logFC > 0 & res_df$FDR < 0.05] <- "Up"
res_df$group[res_df$logFC < 0 & res_df$FDR < 0.05] <- "Down"

tbl <- table(res_df$group)

txt <- paste0(
  "Up = ", tbl["Up"], "\n",
  "Down = ", tbl["Down"], "\n",
  "Not significant = ", tbl["Not significant"]
)

res_df$x = "1"
set.seed(123)
ggplot(res_df, aes(x = x, y = logFC, color = factor(group, levels = c("Up", "Down", "Not significant")))) +
  geom_point(alpha = 0.6, size = 2, position = position_jitter(width = 0.5)) +
  scale_color_manual(name="Type", 
                     values = c("Up" = "red", "Down" = "blue", "Not significant" = "grey"),
                     labels =c("Up" = "Up (n = 3211)", "Down" = "Down (n = 3717)", "Not significant" = "Not significnat (n = 7307)"),
                     guide = guide_legend(nrow=3,
                                          override.aes = list(size = 6))) +
  theme_bw() +
  # annotate("text", x= 1,y=3,   label = txt,
  #          hjust = 0, vjust = 1,
  #          size = 8) +
  labs(x = NULL, y = "Log2 Fold Change", title = "Differential KEGG KOs") +
  #geom_vline(xintercept = c(-1, 1), linetype = "dashed") +
  #geom_hline(yintercept = -log10(0.05), linetype = "dashed") +
  theme(plot.title = element_text(size=30, hjust=0.5),
        axis.title = element_text(size=26),
        axis.text.y = element_text(size=22, color="black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = "bottom",
        legend.title = element_text(size=26),
        legend.text = element_text(size=22))
```


## Fig.S4c Errichenment analysis
```{r, fig.align='center', fig.height=15.5, fig.width=12.0, warning=FALSE, message=FALSE}

de_gene <- list("UP"= substr(res_sig[res_sig$logFC > 0,]$KO,4,9),"DOWN"= substr(res_sig[res_sig$logFC < 0,]$KO,4,9))

download_KEGG("ko")

gene_enrich_result <- compareCluster(
  geneClusters = de_gene,
  fun = "enrichKEGG",
  organism = "ko"

)
gene_enrich_result@compareClusterResult -> test

test <- test %>%
  filter(category != "Human Diseases") %>%
  filter(subcategory != "Cellular community - eukaryotes")

test <- test %>%
  arrange(subcategory, desc(FoldEnrichment))

library(scales)
library(ggh4x)
my_colors <- c(
  "UP" = alpha("#cd1936", 1),
  "DOWN" = alpha("#1445f8", 1)
)

strip_custom <- strip_themed(
  background_x = elem_list_rect(
    fill = my_colors,
    color = "black"
  ),
  text_x = elem_list_text(
    #face = "bold",
    colour = "white"
  )
)

colors <- c(
  "#FFA500", "#8bc24c", "#FFD700", "#2196F3", "#8A4B08",
  "#FF7F00", "#d9534f", "#86519D", "#1B9E77", "#EE4C97",
  "#3B3B98", "#2B5A41", "#E41A1C", "#b4bb72", "#b23256",
  "#377EB8", "#4DAF4A", "#984EA3", "#B22222", "#A65628",
  "#F781BF", "#666666", "#D95F02", "#6495ED", "#E7298A"
)

test$FoldEnrichment[16:17] = mean(test$FoldEnrichment[16:17])/2

saveRDS(test, "E:/HMME/Metagenome/Result_new/bin/bin_dereplecated/kofamscan/up_function.rds")


test1 <- test[test$ID != "ko05111",]
test1[test$ID == "ko02025",]$FoldEnrichment <- mean(test[test$Description == "Biofilm formation",]$FoldEnrichment)

df_scales <- data.frame(
  Panel = c("UP", "DOWM"),
  xmin = c(0, 0),
  xmax = c(4.5,8),
  n = c(4,4)
)
df_scales <- split(df_scales, df_scales$Panel)

scales <- lapply(df_scales, function(x) {
  scale_x_continuous(limits = c(x$xmin, x$xmax), n.breaks = x$n)
})



# add text

p1 <- ggplot(test1, aes(x = abs(FoldEnrichment), y = factor(Description, levels=rev(unique(Description))), fill = subcategory)) +
  facet_grid2(. ~ Cluster, scales = "free_x",
              strip = strip_custom,
              labeller = labeller(Cluster = c("UP"="Increased", "DOWN"="Depleted"))) +
  geom_col(width = 0.7) +
  geom_text(aes(label = GeneRatio), size = 5.5, hjust = -0.2) +
  scale_fill_manual(name = "KEGG Level 2 ", values = colors) +
  scale_y_discrete(position = "right") +
  labs(x="Fold enrichment") +
  xlim(0,8.1) +
  guides(fill = guide_legend(ncol = 2, title.position = "top")) +
  theme_bw() +
  theme(plot.margin = unit(c(1, 1, 1, 1), "cm"),
        plot.title = element_text(color = "black", size = 20, hjust = 0.5),
        axis.title = element_text(color = "black", size = 20), 
        axis.title.y = element_blank(), 
        axis.text.y = element_text(color = "black", size = 16),
        axis.text.x = element_text(color = "black", size = 18, angle = 0, vjust=0.5, hjust=0.5),
        legend.position = "bottom",
        legend.title = element_text(color = "black", size = 20),
        legend.text = element_text(color = "black", size = 16),
        legend.justification = c(0, 1),
        legend.key.spacing.y = unit(0.5,"pt"),
        strip.text = element_text(color = "black",size = 18),
        strip.background =element_rect(fill="lightgrey", size = 0)
  )


print(p1)

# h15.5 w12

```
