---
title: "Fig.S4a & Fig.S5a_PCoA for KEGG and mgPGPT gene composition"
author: "Huanhuan Zhu"
date: "2025-09-28"
output: html_document
---


## Fig.S4a_PCoA for KEGG gene composition 
```{r, fig.align='center', fig.height=6.88, fig.width=6.65, warning=FALSE, message=FALSE}

library(tidyverse)
library(ggplot2)
library(tibble)
library(edgeR)
library(data.table)
library(patchwork)
library(vegan)
library(ape)
library(ggplot2)

rm(list=ls())

setwd("E:/HMME//Metagenome/Result_new/eggnog/")

ko_all <- read.table("tpm/eggnog.summarizeAbundance.KEGG_ko.raw.txt",
                     check.names = F,
                     row.names = 1,
                     header = T)

ko_all <- ko_all[2:nrow(ko_all),]
ko_all_up <- ko_all
ko_all_up <- ko_all_up[,grepl("TS",colnames(ko_all_up))]

count_data <- ko_all_up
count_data <- round(count_data)

col_data <- data.frame(
  sample = colnames(count_data),
  drought = as.numeric(substr(colnames(count_data),1,1))
)
rownames(col_data) <- col_data$sample

ko_t <- t(ko_all_up) 

dist_mat <- vegdist(ko_t, method = "bray")

pcoa_res <- pcoa(dist_mat)

pcoa_df <- data.frame(
  Sample = rownames(pcoa_res$vectors),
  Axis1 = pcoa_res$vectors[,1],
  Axis2 = pcoa_res$vectors[,2]
)

pcoa_df$Generation <- substr(pcoa_df$Sample, 1, 1)
pcoa_df$type <- "KEGG"

# adnois
adonis_res <- adonis2(ko_t ~ pcoa_df$Generation, method = "bray")
adonis_res
perma_label1 <- sprintf("italic(F)*(%.3f * ',' ~ %.3f) == %.3f",
                        adonis_res$Df[1],
                        adonis_res$Df[2],
                        adonis_res$F[1])

perma_label2 <- sprintf("atop(italic(R^2) == %.3f,~ italic(p) < %.3g)",
                        adonis_res$R2[1],
                        adonis_res$`Pr(>F)`[1])

colors <- c("#1b9e77",
            "#3B3B98",
            "#984ea3",
            "#FFD700",
            "#FF7F00",
            "#e41a1c",
            "#8c510a")

p <-  ggplot(pcoa_df, aes(x = Axis1, y = Axis2, fill = Generation)) +
      geom_point(size = 7, shape = 21, alpha = 0.7) +
      scale_fill_manual(values = colors) +
      labs(title = "KEGG") +
      xlab(paste0("PC1 (", round(pcoa_res$values$Relative_eig[1]*100, 1), "%)")) +
      ylab(paste0("PC2 (", round(pcoa_res$values$Relative_eig[2]*100, 1), "%)")) +
      annotate(geom = "text", x = -0.12, y = 0.038, label = perma_label1, parse = TRUE, hjust = -0.00, vjust = -0.1, size = 8) +
      annotate(geom = "text", x = -0.12, y = 0.012, label = perma_label2, parse = TRUE, hjust = -0.00, vjust = -0.1, size = 8) +
      theme_bw() +
      theme(
        legend.position = "bottom",
        plot.title = element_text(size = 30, hjust = 0.5),
        axis.text = element_text(size=22, angle = 0, hjust = 0.5, color = "black"),
        strip.text = element_text(size=26),
        axis.title = element_text(size=26),
        legend.key.height = unit(0.5, "cm"),
        legend.key.width = unit(0.3, "cm"), 
        legend.title = element_text(size = 26), 
        legend.text = element_text(size = 22) 
      )

# h6.88 w6.65
print(p)

```


## Fig.S5a_PCoA for mgPGPT gene composition 
```{r, fig.align='center', fig.height=6.88, fig.width=6.65, warning=FALSE, message=FALSE}

rm(list=ls())

setwd("E:/HMME/Metagenome/Result_new/mgPGPT/")

ko_all <- read.table("mgPGPT.summarizeAbundance.tpm.sseqid.raw.txt",
                     check.names = F,
                     row.names = 1,
                     header = T)

ko_all_up <- ko_all
ko_all_up <- ko_all_up[,grepl("TS",colnames(ko_all_up))]

ko_t <- t(ko_all_up)

dist_mat <- vegdist(ko_t, method = "bray")

pcoa_res <- pcoa(dist_mat)

pcoa_df <- data.frame(
  Sample = rownames(pcoa_res$vectors),
  Axis1 = pcoa_res$vectors[,1],
  Axis2 = pcoa_res$vectors[,2]
)

pcoa_df$Generation <- substr(pcoa_df$Sample, 1, 1)
pcoa_df$type <- "mgPGPT"

adonis_res <- adonis2(ko_t ~ pcoa_df$Generation, method = "bray")
adonis_res

perma_label1 <- sprintf("italic(F)*(%.3f * ',' ~ %.3f) == %.3f",
                       adonis_res$Df[1],
                       adonis_res$Df[2],
                       adonis_res$F[1])

perma_label2 <- sprintf("atop(italic(R^2) == %.3f,~ italic(p) < %.3g)",
                       adonis_res$R2[1],
                       adonis_res$`Pr(>F)`[1])

colors <- c("#1b9e77",
            "#3B3B98",
            "#984ea3",
            "#FFD700",
            "#FF7F00",
            "#e41a1c",
            "#8c510a")

p <-  ggplot(pcoa_df, aes(x = Axis1, y = Axis2, fill = Generation)) +
      geom_point(size = 7, shape = 21, alpha = 0.7) +
      scale_fill_manual(values = colors) +
      labs(title = "mgPGPT") +
      xlab(paste0("PC1 (", round(pcoa_res$values$Relative_eig[1]*100, 1), "%)")) +
      ylab(paste0("PC2 (", round(pcoa_res$values$Relative_eig[2]*100, 1), "%)")) +
      annotate(geom = "text", x = -0.10, y = -0.038, label = perma_label1, parse = TRUE, hjust = -0.00, vjust = -0.1, size = 8) +
      annotate(geom = "text", x = -0.10, y = -0.06, label = perma_label2, parse = TRUE, hjust = -0.00, vjust = -0.1, size = 8) +
      theme_bw() +
      theme(
        legend.position = "bottom",
        plot.title = element_text(size = 30, hjust = 0.5),
        axis.text = element_text(size=22, angle = 0, hjust = 0.5, color = "black"),
        strip.text = element_text(size=26),
        axis.title = element_text(size=26),
        legend.key.height = unit(0.5, "cm"),
        legend.key.width = unit(0.3, "cm"),
        legend.title = element_text(size = 26),
        legend.text = element_text(size = 22) 
      )
# h6.88 w6.65

print(p)

```

